README
Genezzo-Contrib-Clustered 
=========================

Shared data cluster support for Genezzo

Genezzo is an extensible database with SQL and DBI.  It is written in Perl.
Basic routines inside Genezzo are overridden via Havok SysHooks.  Override
routines provide support for shared data clusters.  Routines
provide transactions, distributed locking, undo, and recovery.  

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

To prepare for use of Genezzo::Clustered

  genprepundo.pl

  gendba.pl
  >@havok.sql
  >@syshook.sql
  >@clustered.sql

LIMITATIONS

  This is pre-alpha software; don't use it to store any data you hope
  to see again!

  Transactions, Rollback, etc. are not fully implemented.  Process death
  and necessary cleanup is not detected.

DEPENDENCIES

This module requires these other modules and libraries:

  Genezzo
  FreezeThaw

  OpenDLM

SEE ALSO

  For more information, please visit the Genezzo homepage
  at http://www.genezzo.com

  also 
  http://eric_rollins.home.mindspring.com/genezzo/ClusteredGenezzoDesign.html
  http://eric_rollins.home.mindspring.com/genezzo/cluster.html
  http://opendlm.sourceforge.net/

TODO

  1)   [Genezzo fix done]

  2)   [Genezzo update] Need a exception mechanism when SQL is run to catch
       errors such as running out of disk space.  Currently multi-row
       SQL statements can result in partial results on disk when system
       runs out of disk space.
      
  3)   Use new exception mechanism (above) to handle deadlock with DLM.
       Currently program exits on any error from OpenDLM.  Need to modify
       inline::C interface to provide deadlock return code distinct from
       other errors.  Need to modify GLock Perl code to throw SQL exception.

  4)   [Genezzo update done]

  5)   done.

  6)   [Genezzo update] Need way to invalidate entire buffer cache.
       Simply release all blocks, don't write them to disk (disk write
       should already have been done in commit case). 

  7)   Invalidate buffer cache (using new mechanism above) before releasing 
       all locks on commit or rollback.  Actually only invalidate 
       non-system tablespace portion, see below.

  8)   [Genezzo update done] 

  9)   done

  10)  [Genezzo fix] fix "mystery" writes.  These bogus writes create 
       unnecessary write locks,  generate unnecessary undo, and confuse 
       rollback code.  Solution may be to attach syshook to new attachment 
       point instead of Genezzo::BufCa::DirtyScalar::STORE.  Yes, some of
       the writes are not "mysteries".  They occur when a pre-existing
       buffer cache block is reused -- the reloading of new data from disk
       or memory into the block triggers the STORE callback.  The cure
       is to attach the syshook in a more intelligent location that is
       aware of the different cases.

  11)  [Genezzo update] complete tablespace support.

  12)  Use tablespace support (above) to restrict locking to non-system
       tablespace tables.  Otherwise we need to lock additional blocks
       (for system tables) already read at startup prior to syshook
       initialization.  And locking system tables prevents other
       instances from running in parallel.  [this may not solve problems...]

  13)  [Genezzo update] add pre-havok command line param which specifies
       routine to run prior to opening database.

  14)  Use pre-havok (above) to perform recovery prior to opening database.
       Havok/syshook routines won't run if database cannot be opened due
       to corruption.

  13)  [Genezzo redesign] Storing table directory and free space list in
       block zero of each tablespace prevents any concurrency of writes
       with reads.  Also cached structures become invalid whenever block
       zero is written.

  14)  [Genezzo refactoring] Move Verify and Update checksum routines
       out of Genezzo::BufCA::BCFile::_filereadblock and _filewriteblock
       and into Util.

  15)  Modify VerifyChecksum and UpdateChecksum to use Util routines.

  99)  After the above, complete rest of development per Design Document:
   a)  [Done] Detect blocks needing recovery via process_id in block.
   b)  [Done] Recover dead processes
   c)  etc...

  100) Recognize incomplete undo blocks and read buddy block instead.

COPYRIGHT AND LICENCE

    Copyright (C) 2005 by Eric Rollins.  All rights reserved.

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

Address bug reports and comments to rollins@acm.org


Changes
Revision history for Perl extension Genezzo::Contrib::Clustered.
0.33  Wed Jan 18 23:00:00 2006
	Revised POD in ModPerlWrap.  Requires Apache 2.  Signal handling
	  problem on Mac OS X is due to Apache 1.3.

0.32  Mon Jan 16 23:00:00 2006
	Revised SIGUSR2 handling in GLockUR and ModPerlWrap.
	OS X 10.3.9 kill SIGUSR2 handling is sometimes delayed till 
	  next request (WHY?).

0.31  Sun Jan 15 01:00:00 2006
	POD doc fixes.

0.30  Sat Jan 14 23:00:00 2006
	Added new locking module GLockUR, based on Unix Record locking.
	Works on OS X, EXCEPT processes exit immediately after 
	  signal handlers are run (why?)
	Also on OS X need to add to httpd.conf
	  PerlHeaderParserHandler "sub { tie *STDOUT, 'Apache' unless tied *STDOUT; }"

0.29  Tue Jan 10 20:30:00 2006
	Fix race condition in ModPerlWrap.

0.28  Tue Jan 03 23:30:00 2006
	Modify ModPerlWrap to run better when preloaded with PerlModule.
	Added use test case for ModPerlWrap.

0.27  Tue Jan 03 09:00:00 2006
	Modify ModPerlWrap to reset $processing and $restart on each
	  PageStart()

0.26  Mon Jan 02 23:00:00 2006
	Modified VerifyChecksum and UpdateChecksum to use new
	  Genezzo::Block::Util routines.

0.25  Sun Jan 01 17:00:00 2006
	Added ModPerlWrap.pm and sample genezzo_form.pl to use Apache
	  web server as multi-user XML over HTTP interface to 
	  Clustered Genezzo.
	Added support for asyncronous notification of blocking lock
	  requests.  See ModPerlWrap or modified genloopclient.pl for
	  examples.

0.24  Sat Dec 24 01:00:00 2005
	Modified DLM code to raise DEADLOCK error on deadlock.
	Added experimental Kngz_Execute hook to catch DEADLOCK.
	Modified DLM code to support demoteAll (to SH) besides unlockAll.
	Commit now uses demoteAll insead of unlockAll, to match current
	  buffer cache behavior.
	Added DLM function ast_poll to determine whether there are any
	  recent requests for locks held by that process.
	Added new client genloopclient.pl to test locking interactions between 
	  clients. 
	Now recommend rollback as first action in gendba or scripts,
	  so all blocks will be locked.

0.23  Fri Dec 23 12:00:00 2005
	Modify ReadBlock to replace instead of wrap _filereadblock
	  to avoid duplicate reads.
	Attempted moving DirtyBlock to BufCaElt::_dirty; doesn't work
	  so left old DirtyScalar::STORE method registered.
	sync is no longer called on non-Cygwin Win32.  autoflush(1) is
	  called instead.  Yes, this probably gives terrible performance,
	  but it should pass CPAN Testers :-)

0.22  Tue Dec 20 11:00:00 2005
	HTML doc fix.

0.21  Mon Dec 19 23:00:00 2005
	Handle corrupted transaction code byte case.
	Add another TODO.

0.20  Mon Dec 19 07:00:00 2005
	Update ClusteredGenezzoDesign.html
	Pad transaction status block with '=' instead of '-'
	Restore write of PID in transaction status block
	Update Pod

0.19  Sun Dec 18 23:00:00 2005
	Detect blocks needing recovery via PID in block.
	Recover blocks and processes, resetting PID to zero.
	Converted entire package to OO methods.

0.18  Sun Oct 23 19:00:00 2005
        Writing PIDs to blocks.  No set to zero yet.
	Print to STDERR, not STDOUT.

0.17  Tue Aug 30 11:59:00 2005
        modified PrepUndo to use DictSetFileInfo to add undo_filename
	  to default.dbf file header
	modified clustered.sql to use oo_require
	modified Clustered to use new OO style methods, use dict
	  to look up undo_filename and gnz_home

0.16  Sun Aug 21 12:40:00 2005
        added TODO to README

0.15  Sun Aug 21 10:21:00 2005
	moved body of genprepundo.pl to PrepUndo.pm
	expanded tests to verify rollback really works

0.14  Wed Aug 10 21:40:00 2005
	maintain own dirty_block list instead of relying on dirty flag
	pod fixes
	add license to Makefile.PL

0.12  Thu Aug 04 07:30:00 2005
	pod fixes

0.11  Thu Aug 04 00:20:00 2005
	pod fixes

0.10  Wed Aug 03 19:45:00 2005
	use SVR lock to allocate unique processid at startup
	fix strictness of genprepundo.pl
	variable renamings
	pod fixes
	documentation fixes

0.09  Mon Aug 01 11:55:00 2005
	documentation fixes

0.08  Mon Aug 01 08:00:00 2005
	Added 1; to GTXLock
	documentation fixes

0.07  Sun Jul 31 23:00:00 2005
	Added OpenDLM-based locking support

0.06  Sun Jul 31 20:45:00 2005
	added ClusteredGenezzoDesign.html design document
	added message to genprepundo.pl

0.05  Thu Jul 28 22:00:00 2005
	fix make test
	documentation additions

0.04  Tue Jul 26 23:00:00 2005
        transactional fixes
	documentation fixes

0.03  Mon Jul 25 21:00:00 2005
	documentation fixes
	raw device fix
	documented params for genprepundo.pl

0.02  Mon Jul 25 01:00:00 2005
	documentation fixes
	raw device fix

0.01  Mon Jul 25 00:12:16 2005
	- original version; created by h2xs 1.23 with options
		-X Genezzo::Contrib::Clustered



lib/Genezzo/Contrib/Clustered/PrepUndo.pm
#!/usr/bin/perl
#
# PrepUndo.pm
#
# Initialize undo file for Clustered Genezzo
#
#
package Genezzo::Contrib::Clustered::PrepUndo;

use strict;
use warnings;
use Genezzo::GenDBI;
use Data::Dumper;
use Carp;
use Genezzo::Dict;
use Genezzo::Block::RDBlock;
use Genezzo::Block::Std;
use FreezeThaw;
use Genezzo::Util;
use Genezzo::Contrib::Clustered::Clustered;

our $GZERR = sub {
    my %args = (@_);

    return 
        unless (exists($args{msg}));

    my $warn = 0;
    if (exists($args{severity}))
    {
        my $sev = uc($args{severity});
        $sev = 'WARNING'
            if ($sev =~ m/warn/i);

        # don't print 'INFO' prefix
        if ($args{severity} !~ m/info/i)
        {
            printf ("%s: ", $sev);
            $warn = 1;
        }

    }
    print $args{msg};
    # add a newline if necessary
    print "\n" unless $args{msg}=~/\n$/;
#    carp $args{msg}
#      if (warnings::enabled() && $warn);
    
};

sub MakeSQL
{
    my $bigSQL; 
    ($bigSQL = <<EOF_SQL) =~ s/^\#//gm;
#REM prepare database for Genezzo::Contrib::Clustered
#REM ATAThavok.sql and ATATsyshook.sql first
#insert into sys_hook (xid, pkg, hook, replace, xtype, xname, args, owner, creationdate, version) values (1000, 'Genezzo::BufCa::BCFile', '_filereadblock', 'ReadBlock_Hook', 'oo_require', 'Genezzo::Contrib::Clustered::Clustered', 'ReadBlock', 'SYSTEM', TODAY, '1');
#insert into sys_hook (xid, pkg, hook, replace, xtype, xname, args, owner, creationdate, version) values (1001, 'Genezzo::BufCa::DirtyScalar', 'STORE', 'DirtyBlock_Hook', 'oo_require', 'Genezzo::Contrib::Clustered::Clustered', 'DirtyBlock', 'SYSTEM', TODAY, '1');
#insert into sys_hook (xid, pkg, hook, replace, xtype, xname, args, owner, creationdate, version) values (1002, 'Genezzo::GenDBI', 'Kgnz_Commit', 'Commit_Hook', 'oo_require', 'Genezzo::Contrib::Clustered::Clustered', 'Commit', 'SYSTEM', TODAY, '1');
#insert into sys_hook (xid, pkg, hook, replace, xtype, xname, args, owner, creationdate, version) values (1003, 'Genezzo::GenDBI', 'Kgnz_Rollback', 'Rollback_Hook', 'oo_require', 'Genezzo::Contrib::Clustered::Clustered', 'Rollback', 'SYSTEM', TODAY, '1');
#insert into sys_hook (xid, pkg, hook, replace, xtype, xname, args, owner, creationdate, version) values (1003, 'Genezzo::GenDBI', 'Kgnz_Execute', 'Execute_Hook', 'oo_require', 'Genezzo::Contrib::Clustered::Clustered', 'Execute', 'SYSTEM', TODAY, '1');
#insert into sys_hook (xid, pkg, hook, replace, xtype, xname, args, owner, creationdate, version) values (1004, 'Genezzo::BufCa::BCFile', '_init_filewriteblock', '_init_fwb_Hook', 'oo_require', 'Genezzo::Contrib::Clustered::Clustered', '_init_filewriteblock', 'SYSTEM', TODAY, '1');
#commit;
#shutdown;
#REM restart gendba.pl from command line, so havok routines will be redefined
#quit;
EOF_SQL
    my $now = Genezzo::Dict::time_iso8601();
    $bigSQL =~ s/ATAT/\@/gm;
    $bigSQL =~ s/TODAY/\'$now\'/gm;
    my $pver = $Genezzo::Contrib::Clustered::Clustered::VERSION;
    $bigSQL =~ s/GPU_VERSION/$pver/gm;
    $bigSQL = "REM Generated by " . __PACKAGE__ . " version " .
        $pver . " on $now\nREM\n" . $bigSQL;

#    print $bigSQL;

    return $bigSQL;
}


my $glob_init;
my $glob_gnz_home;
my $glob_shutdown; 
my $glob_id;
my $glob_defs;

sub prepareUndo
{
    #my $self = shift;
    my %optional = (
                    number_of_processes => 256,
                    undo_blocks_per_process => 220,
		    undo_filename => "undo.und"
                    );

    my %required = (
                    );

    my %args = (%optional,
                @_);

    return 0
        unless (Validate(\%args, \%required));


    my $glob_undo_filename = (exists($args{undo_filename}))
        && (defined($args{undo_filename}))
        && (length($args{undo_filename})) ? $args{undo_filename} : "";

    my $glob_gnz_home = $args{gnz_home};
    my $glob_procs = $args{number_of_processes};  # num of processes supported
    # undo blocks per process
    my $glob_blocks_per_proc = $args{undo_blocks_per_process};

    my $dbh = Genezzo::GenDBI->connect($glob_gnz_home,
                                   "NOUSER",
                                   "NOPASSWORD",
                                   {GZERR => $GZERR});

    if(!defined($dbh)){
        Carp::croak("failed connect");
      }

    $dbh->do("startup"); # start the database
    
#-----------------------------------------------------------------
# locate home
    my $stmt =    "select pref_value from _pref1 where pref_key='home'";
    
    my $sth = $dbh->prepare($stmt);
    
    if(!defined($sth)){
        Carp::croak("failed prepare 2");
      }
    
    my $ret = $sth->execute();
    
    if(!defined($ret)){
        Carp::croak("failed execute 2");
      }
    
    my $ggg = $sth->fetchrow_hashref();
    
    if(!defined($ggg)){
        Carp::croak("zero rows 2");
      }
    
    my $pref_home = $ggg->{pref_value};
    
    if ($glob_undo_filename eq ""){
        if($pref_home eq "/dev/raw"){
            print "ERROR:  raw device -undo_filename required\n";
            exit();
        }else{
            $glob_undo_filename = "undo.und";
        }
    }
    
#-----------------------------------------------------------------
# read per-file info
    $stmt = "select fileidx,filename,blocksize,numblocks,tsid from _tsfiles";
    
    $sth = $dbh->prepare($stmt);
    
    if(!defined($sth)){
        Carp::croak("failed prepare");
      }
    
    $ret = $sth->execute();
    
    if(!defined($ret)){
        Carp::croak("failed execute");
      }
    
    my $undoHeader = {
        "procs" => $glob_procs,
        "blocks_per_proc" => $glob_blocks_per_proc,
        "files" => {}
    };
    
    while(1)
    {
        my $ggg = $sth->fetchrow_hashref();
        
        last
            unless(defined($ggg));
        
        # look up header size in file FIXME
        my $full_filename;
        
        if($pref_home eq "/dev/raw"){
            $full_filename = "$pref_home/$ggg->{filename}";
        }else{
            $full_filename = "$pref_home/ts/$ggg->{filename}";
        }
        
        $ggg->{full_filename} = $full_filename;
        
        my $fh;
        open($fh, "<$full_filename")
            or die "open $full_filename failed: $!\n";
        
        my ($hdrsize, $version, $blocksize, $h1) = 
            Genezzo::Util::FileGetHeaderInfo(filehandle => $fh, 
	        filename => $full_filename);
        
        $ggg->{hdrsize} = $hdrsize;
        
        $undoHeader->{files}->{$ggg->{fileidx}} = $ggg;
    }
    
#-----------------------------------------------------------------
    my $dictobj = $dbh->{dictobj};
    $dictobj->DictSetFileInfo(newkey => "undo_filename",
        newval => $glob_undo_filename);

    $dbh->do("commit");
    
    my $full_filename;
    
    if($pref_home eq "/dev/raw"){
        $full_filename = "$pref_home/$glob_undo_filename";
    }else{
        $full_filename = "$pref_home/ts/$glob_undo_filename";
    }
    
#-----------------------------------------------------------------
# store per-file info in file header block
    my $frozen_undoHeader = FreezeThaw::freeze $undoHeader;
    
# construct an empty byte buffer
    my $blocksize = $Genezzo::Block::Std::DEFBLOCKSIZE; 
    my $buff = "\0" x $blocksize;
    
    my %tied_hash = ();
    
    my $tie_val = 
        tie %tied_hash, 'Genezzo::Block::RDBlock', (refbufstr => \$buff,
						    blocksize => 
						    $blocksize);
    
    my $newkey = $tie_val->HPush($frozen_undoHeader);
    
    my $fh;
    open($fh, ">$full_filename")
        or die "open $full_filename failed: $!\n";
    
    gnz_write ($fh, $buff, $blocksize)
        == $blocksize
        or die "bad write - file $full_filename : $! \n";
    
#-----------------------------------------------------------------
# mark process status block for each process as no outstanding transaction
    my $i;
    for($i = 0; $i < $glob_procs; $i++){
        $buff = "-" x 10;
        my $procstr = sprintf("%10d", $i);
        $buff = $buff . $procstr;
        $buff = $buff . ( "=" x ($blocksize - 20) );
        
        gnz_write ($fh, $buff, $blocksize)
            == $blocksize
            or die "bad write - file $full_filename ($i): $! \n";
    }
    
    close $fh;
    
    print "finished initialization of $full_filename.\n" ;

}

1;

__END__
# Below is stub documentation for your module. You better edit it!

=head1 NAME

Genezzo::Contrib::Clustered::PrepUndo - Prepare undo file for Clustered Genezzo

=head1 SYNOPSIS

    prepareUndo(gnz_home => "/dev/raw", undo_filename => "raw2");

=head1 DESCRIPTION

  Creates or re-initializes undo file under GNZ_HOME.  By default
  file is named undo.und.  File header contains basic information about
  all other files in Genezzo installation.  genprepundo.pl must
  be run whenever a new file is added to the Genezzo installation.

  Undo file format is documented in Clustered.pm.

=head1 FUNCTIONS

=over 4

=item prepareUndo(gnz_home => GNZ_HOME, undo_filename => UNDO_FILENAME,
    number_of_processes => NUMBER_OF_PROCESSES,
    undo_blocks_per_process => UNDO_BLOCKS_PER_PROCESS )

=item B<gnz_home>
 
    Supply the location for the gnz_home installation.  If 
    specified, it overrides the GNZ_HOME environment variable.

=item B<undo_filename>

    Supply the name of the undo file.  If not specified, it
    defaults to undo.und for file system devices.  It must
    be specified for raw devices.

=item B<number_of_processes>

    Specify the maximum number of processes supported by undo.
    If not specified defaults to 256.

=item B<undo_blocks_per_process>

    Specify the number of blocks (which list fileno-blockno pairs)
    for a process.  If not specified defaults to 220.

=back

=head1 AUTHOR

  Eric Rollins, rollins@acm.org

=head1 COPYRIGHT AND LICENSE

  Copyright (c) 2005 Eric Rollins.  All rights reserved.

  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  any later version.


  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

  Address bug reports and comments to rollins@acm.org

=cut
